### PR TITLE
#38: The Heartbeat Update

### PR DESCRIPTION
**Pull request by itsthejoker**
Adds in a new dependency, CherryPy, and enables the `build_bot` function to also create a small  webserver with one API JSON endpoint to serve as a healthcheck. This information is saved to Redis and will be used by a standalone application to perform healthchecks on all the bots.

Usage in terminal (with Redis running):
```
>>> from tor_core.initialize import build_bot
>>> build_bot('bot_archiver', '9000', full_name='the_archivist')
Version 4.4.0 of praw is outdated. Version 4.5.1 was released Monday May 08, 2017.
[2017-07-04 00:10:40,017] - [configure_logging] - Not running with Bugsnag!
[2017-07-04 00:10:40,017] - [log_header] - **************************************************
[2017-07-04 00:10:40,017] - [log_header] - Starting!
[2017-07-04 00:10:40,017] - [log_header] - **************************************************
[2017-07-04 00:10:51,439] - [configure_heartbeat] - Heartbeat port: 40101
[2017-07-04 00:10:51,543] - [start_heartbeat_server] - Cherrypy heartbeat started!
[2017-07-04 00:10:51,544] - [build_bot] - Bot built and initialized!
>>>
```
In a separate terminal:
```python
>>> import requests
>>> requests.get('http://localhost:40101').json()
```
You'll get back:
```json
{"bot_name": "the_archivist", "bot_version": "9000", "core_version": "0.2.0"}
```
In the original terminal:
```
>>> from tor_core.helpers import stop_heartbeat
>>> from tor_core.config import config
>>> stop_heartbeat(config)
[2017-07-04 00:11:02,418] - [stop_heartbeat] - Stopped heartbeat!
```
`stop_heartbeat` is called automatically when using `run_until_dead`, and that makes the whole process automatic.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -9,3 +9,6 @@ praw.ini
 .cache/
 .idea/
 __pycache__/
+heartbeat.port
+slack.key
+bugsnag.key
```
Comment (TheLonelyGhost) [inline .gitignore]: What about just `*.key` here?

File: CHANGELOG.md [hunk 2]
```diff
@@ -3,6 +3,10 @@
 We follow [Semantic Versioning](http://semver.org/) as a way of measuring stability of an update. This
 means we will never make a backwards-incompatible change within a major version of the project.
 
+## v0.2.0
+
+- added Cherrypy server hooks into `build_bot()` and `run_until_dead()` to create a json endpoint for use as a heartbeat check (by @itsthejoker)
```
Comment (TheLonelyGhost) [inline CHANGELOG.md]: What about shortening this line to something like:

> Added heartbeat web service to check status of a bot (by @itsthejoker)

The changelog is intended to be very high level summary and what is currently written seems like there's too much technical detail, a problem I have all the time. Or is there are reason for adding that detail?

File: tor_core/__init__.py [hunk 3]
```diff
@@ -1 +1 @@
-__version__ = '0.1.0'
+__version__ = '0.2.0'
```
Comment (TheLonelyGhost) [inline tor_core/__init__.py:1]: 👍

File: tor_core/config.py [hunk 4]
```diff
@@ -156,6 +156,9 @@ class Config:
 
 # ----- Compatibility -----
 config = Config
+config.name = None
+config.bot_version = 0  # this should get overwritten by the bot process
```
Comment (TheLonelyGhost) [inline tor_core/config.py]: Good use of commenting here.

For bot versions, though, are we trying to limit them to integers? If not, having the `0` as a string might work better.

File: tor_core/config.py [hunk 5]
```diff
@@ -175,3 +178,6 @@ class Config:
 
 # section for gifs
 config.no_gifs = []
+
+# enables debug information for the cherrypy heartbeat server
+config.heartbeat_logging = False
```
Comment (TheLonelyGhost) [inline tor_core/config.py:183]: I never thought of giving the option to log or not just for the heartbeat server. Good idea! 👍

File: tor_core/heartbeat.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    for the heartbeat. Usage:
+
+    >>> import requests
+    >>> requests.get('http://localhost:{portnumber}').json()
+    returns:
+    {
+        'bot_name': 'this_is_an_awesome_bot',
+        'bot_version': '9001',
+        'core_version': '0.2.0'
+    }
```
Comment (TheLonelyGhost) [inline tor_core/heartbeat.py:28]: You know, I bet if we changed this around a bit we might eventually be able to use this for a doctest style of testing... 😁 

That's a bit out of scope for this PR though. Nice work! 👍

File: tor_core/heartbeat.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+    # update the global config (separate from the application config above)
+    cherrypy.config.update(
+        {
+            'server.socket_port': config.heartbeat_port
+        }
+    )
+    logging.info('Heartbeat port: {}'.format(config.heartbeat_port))
+
+    if config.heartbeat_logging is False:  # defaults to false
```
Comment (TheLonelyGhost) [inline tor_core/heartbeat.py:42]: Since we don't realistically need to care about the datatype of this config value, what about opening this up to any false-y value?

```python
if not config.heartbeat_logging:
    # [code]
```

File: tor_core/helpers.py [hunk 8]
```diff
@@ -75,7 +76,7 @@ def send_to_slack(message, config):
     return
 
 
-def explode_gracefully(error, config, bot_name=None):
+def explode_gracefully(error, config):
```
Comment (TheLonelyGhost) [inline tor_core/helpers.py:79]: Nice! This definitely seems like a better place to store the bot name. 👍

File: tor_core/helpers.py [hunk 9]
```diff
@@ -179,7 +179,53 @@ def update_wiki_page(pagename, content, config):
...
+    the command line.
+
+    :param port: int, the port number
+    :return: None
+    """
+    config.redis.srem('active_heartbeat_ports', port)
+    logging.info('Removed port from set of heartbeats.')
+
+
+def stop_heartbeat(config):
```
Comment (TheLonelyGhost) [inline tor_core/helpers.py:196]: This is a morbid name... I love it. 👍

File: tor_core/helpers.py [hunk 10]
```diff
@@ -179,7 +179,53 @@ def update_wiki_page(pagename, content, config):
...
+
+
+def run_until_dead(func, exceptions=default_exceptions):
+    """
+    The official method that replaces all that ugly boilerplate required to
+    start up a bot under the TranscribersOfReddit umbrella. This method handles
+    communication issues with Reddit, timeouts, and handles CTRL+C and
+    unexpected crashes.
+
+    :param func: the function that you want to run; this will automatically be
```
Comment (TheLonelyGhost) [inline tor_core/helpers.py]: Did you mean to start this param description with a capital? `the function` -> `The function`?

File: tor_core/helpers.py [hunk 11]
```diff
@@ -179,7 +179,53 @@ def update_wiki_page(pagename, content, config):
...
+        passed the config object. Historically, this is the only thing needed
+        to start a bot.
+    :param exceptions: A tuple of exception classes to guard against. These are
+        the praw connection exceptions by default; things like timeouts and
+        general connection errors, but they can be overridden with your own
+        set of exceptions if you feel it necessary.
```
Comment (TheLonelyGhost) [inline tor_core/helpers.py]: Seems pretty wordy for a description of the parameter. What about something more concise, such as:

> A tuple of exceptions to guard against (default: certain PRAW connection exceptions)

File: tor_core/helpers.py [hunk 12]
```diff
@@ -179,7 +179,53 @@ def update_wiki_page(pagename, content, config):
...
+    :return: None. When this returns, everything necessary has been processed
+        to allow the bot to die. Set `run_until_dead()` as the last line under
+        if __name__ == '__main__' and the bot will exit cleanly after this
+        returns.
```
Comment (TheLonelyGhost) [inline tor_core/helpers.py]: Not sure we need this much documentation on just the return value.

What if we put the `if __name__ == '__main__'` stuff in the main doc above?

File: tor_core/helpers.py [hunk 13]
```diff
@@ -193,7 +239,9 @@ def run_until_dead(func, bot_name=None, exceptions=default_exceptions):
 
     except KeyboardInterrupt:
         logging.info('User triggered shutdown. Shutting down.')
+        stop_heartbeat(config)
         sys.exit(0)
 
     except Exception as e:
-        explode_gracefully(e, config, bot_name=bot_name)
+        stop_heartbeat(config)
+        explode_gracefully(e, config)
```
Comment (TheLonelyGhost) [inline tor_core/helpers.py:243]: 🤣  Still love that naming... 👍

File: tor_core/initialize.py [hunk 14]
```diff
@@ -211,17 +213,67 @@ def initialize(config):
...
+    Attempts to pull an existing port number from the filesystem, and if it
+    doesn't find one then it generates the port number and saves it to a key
+    file.
+
+    :param config: the global config object
+    :return: int; the port number to use.
+    """
+    try:
+        # have we already reserved a port for this process?
+        port = int(open('heartbeat.port', 'r').readline().strip())
```
Comment (TheLonelyGhost) [inline tor_core/initialize.py]: Not quite sure how Python handles this form of `open()`, but it looks like it might leave an open file descriptor which is only closed when the program does some garbage collection or exits. ([see also](https://stackoverflow.com/questions/9282967/how-to-open-a-file-using-the-open-with-statement))

It might be safer to do something like this:

```python
with open('heartbeat.port', 'r') as port_file:
    port = int(port_file.readline().strip())
```

Otherwise, if you don't use the `with` keyword, you would have to call `close()` on the variable returned from `open()` in order to close the file descriptor again.

File: tor_core/initialize.py [hunk 15]
```diff
@@ -211,17 +213,67 @@ def initialize(config):
...
+        logging.debug('Found existing port saved on disk')
+        return port
+    except OSError:
+        pass
+
+    while True:
+        port = random.randrange(40000, 40200)  # is 200 ports too much?
```
Comment (TheLonelyGhost) [inline tor_core/initialize.py:235]: Not too many at all. I'm sure if someone is building over 200 bots on one server, there are already some issues with scaling that they would need to address first.

File: tor_core/initialize.py [hunk 16]
```diff
@@ -211,17 +213,67 @@ def initialize(config):
...
+        if config.redis.sismember('active_heartbeat_ports', port) == 0:
+            config.redis.sadd('active_heartbeat_ports', port)
+
+            # create that file we looked for earlier
+            port_file = open('heartbeat.port', 'w')
+            port_file.write(str(port))
+            port_file.close()
```
Comment (TheLonelyGhost) [inline tor_core/initialize.py]: What about shortening this using `with`? Then, in case you do more while the file is open, the `port_file.close()` line going missing won't create something like a memory leak.

```python
with open('heartbeat.port', 'w') as port_file:
    port_file.write(str(port))
```

File: tor_core/initialize.py [hunk 17]
```diff
@@ -211,17 +213,67 @@ def initialize(config):
...
+
+
+def build_bot(
+    name,
+    version,
+    full_name=None,
+    log_name='transcribersofreddit.log',
+    require_redis=True,
+    heartbeat_logging=False
+):
```
Comment (TheLonelyGhost) [inline tor_core/initialize.py:254]: Given how many params are here and have defaults, maybe this is a candidate for `**kwargs`?

File: tor_core/initialize.py [hunk 18]
```diff
@@ -211,17 +213,67 @@ def initialize(config):
...
+        match the settings in praw.ini
+    :param version: string; the version number for the current bot being run
+    :param full_name: string; the descriptive name of the current bot being
+        run; this is used for the heartbeat and status
+    :param log_name: string; the name to be used for the log file on disk. No
+        spaces.
+    :param require_redis: bool; triggers the creation of the Redis instance.
+        Any bot that does not require use of Redis can set this to False and
+        not have it crash on start because Redis isn't running.
+    :return: None
```
Comment (TheLonelyGhost) [inline tor_core/initialize.py:269]: Given what I see below, isn't the return actually void instead of `None`?

**Review by TheLonelyGhost — CHANGES_REQUESTED** [review]
A lot going on here, but some really neat concepts and some good progress. Nice job!

File: CHANGELOG.md (hunk 2 above)
Comment (itsthejoker) [reply CHANGELOG.md]: No specific reason. I'll use your suggestion.

File: tor_core/config.py (hunk 4 above)
Comment (itsthejoker) [reply tor_core/config.py]: Because this is in the core, it just gets overwritten with whatever gets passed in, so type doesn't matter. That being said, it would probably be clearer if I replaced it with `'0.0.0'`.

File: tor_core/config.py (hunk 5 above)
Comment (itsthejoker) [reply tor_core/config.py:183]: It's an option from necessity! :smile: Cherrypy is very verbose, so we need to be able to turn off the logging so that it doesn't clutter up everything else.

File: tor_core/heartbeat.py (hunk 7 above)
Comment (itsthejoker) [reply tor_core/heartbeat.py:42]: I perceive that `if config.heartbeat_logging is False` is clearer; it's not clear what `config.heartbeat_logging` is otherwise, even if it's not really important.

File: tor_core/helpers.py (hunk 11 above)
Comment (itsthejoker) [reply tor_core/helpers.py]: My personal stance is that documentation basically can't be too wordy -- that being said, I'll trim this one down a little but I'd like to keep the meat of it there.

File: tor_core/helpers.py (hunk 12 above)
Comment (itsthejoker) [reply tor_core/helpers.py]: Yeah, I can definitely move it to the readme.

File: tor_core/initialize.py (hunk 14 above)
Comment (itsthejoker) [reply tor_core/initialize.py]: My understanding is that chaining the operations onto the end implicitly closes it, but using the `with` handler isn't a bad idea.

File: tor_core/initialize.py (hunk 17 above)
Comment (itsthejoker) [reply tor_core/initialize.py:254]: Default kwargs still have to be defined, so setting `**kwargs` after would actually make it longer.

File: tor_core/initialize.py (hunk 18 above)
Comment (itsthejoker) [reply tor_core/initialize.py:269]: Python's `'void'` return is `None`.

File: tor_core/heartbeat.py (hunk 7 above)
Comment (TheLonelyGhost) [reply tor_core/heartbeat.py:42]: That's fair, I can see how `x is False` would be clearer to indicate a datatype. I could go on about variable naming indicating type, but at the end of the day it's just one variable in one small scope, so it doesn't matter that much. 👍

File: tor_core/helpers.py (hunk 11 above)
Comment (TheLonelyGhost) [reply tor_core/helpers.py]: I'm of a stance where I obsess over any narrative or docs I write. If fewer words with the same intended meaning, that option is preferable.

I just happen to have a lot of experience with being overloaded with information. At times it's difficult to separate what's pertinent, so less is more from what I've seen.

File: tor_core/initialize.py (hunk 18 above)
Comment (TheLonelyGhost) [reply tor_core/initialize.py:269]: The type is a null-type, a.k.a. `None`, and I get that. I've also seen docs where, if the return type holds no meaning, the documentor notes a "void" return type. That's the case in YARDoc (Ruby) anyway.

From [this stack overflow post](https://stackoverflow.com/a/36797437/1236035), it looks like it's customary to document the actual type that's returned (`None`) rather than the intention (`void`). Carry on! 👍

File: tor_core/initialize.py (hunk 17 above)
Comment (TheLonelyGhost) [reply tor_core/initialize.py:254]: That's a valid argument. 👍 

*snicker* Valid... Argument... get it? 😆

File: CHANGELOG.md [hunk 19]
```diff
@@ -5,9 +5,9 @@ means we will never make a backwards-incompatible change within a major version
...
-- added Cherrypy server hooks into `build_bot()` and `run_until_dead()` to create a json endpoint for use as a heartbeat check (by @itsthejoker)
+- Added heartbeat web service to check status of a bot (by @itsthejoker)
 
 ## v0.1.0
 
-- created first version of tor_core from parts of u/ToR (by @itsthejoker)
-- added in basic testing suite
\ No newline at end of file
+- Created first version of tor_core from parts of u/ToR (by @itsthejoker)
+- Added in basic testing suite
```
Comment (TheLonelyGhost) [inline CHANGELOG.md:13]: Caps is nice. 👍

File: README.md [hunk 20]
```diff
@@ -2,6 +2,14 @@
...
 Everything a growing bot needs!
 
+The important parts are two functions that assist with setting up and running the bot.
+
+- `tor_core.initialize.build_bot()` creates the global config object, populates it with settings from Reddit, and also puts together all the local database connections and whatnot.
+
+- `tor_core.helpers.run_until_dead()` does what it says; the official method that replaces all that ugly boilerplate required to start up a bot under the TranscribersOfReddit umbrella. This method handles communication issues with Reddit, timeouts, and handles CTRL+C and unexpected crashes.
+
+When initializing the new bot, all you have to do is call `build_bot()` and then pass your runtime function into `run_until_dead()`. When `run_until_dead()` returns, everything necessary has been processed to allow the bot to die. Set `run_until_dead()` as the last line under `if __name__ == '__main__'` and the bot will exit cleanly after this returns.
+
```
Comment (TheLonelyGhost) [inline README.md:12]: Ah, I was originally thinking putting it higher in the docstring, like before the param-specific notes. I like this much better! Nice work!

**Review by TheLonelyGhost — APPROVED** [review]
Looks good to me! 👍
